### PR TITLE
Pass CFileItem ByConstRef in CDVDFactoryInputStream::CreateInputStream()

### PR DESCRIPTION
**Pull request by ghost**
Avoid copy of CFileItem (its really a monster struct with all its members)

### CODE REVIEW

File: xbmc/cores/VideoPlayer/DVDInputStreams/DVDFactoryInputStream.cpp [hunk 1]
```diff
@@ -137,12 +137,15 @@ CDVDInputStream* CDVDFactoryInputStream::CreateInputStream(IVideoPlayer* pPlayer
...
 
+    //why are members set ath this global place? Disabled (mapfau)
+    /*
     if (fileitem.ContentLookup())
     {
       // request header
       fileitem.SetMimeType("");
       fileitem.FillInMimeType();
     }
+    */
```
Comment (mapfau) [inline xbmc/cores/VideoPlayer/DVDInputStreams/DVDFactoryInputStream.cpp]: the code above seems not to be at the right place, should be moved to the implementation class where its really thought for.

**Comment by FernetMenta** [comment]
this won't work. you crate and DVDInputStream object and pass a reference to an object that may only be temporary and long gone while DVDInputStream is still in use.

File: xbmc/cores/VideoPlayer/DVDInputStreams/DVDFactoryInputStream.cpp (hunk 1 above)
Comment (FernetMenta) [inline xbmc/cores/VideoPlayer/DVDInputStreams/DVDFactoryInputStream.cpp]: this breaks behaviour, see the following lines

**Comment by mapfau** [comment]
the ref is copied in all constructors to a member. This should work.

**Comment by FernetMenta** [comment]
I correct myself. The inputStream objects copy the fileItem anyways. So it is ok to pass by ref.

**Comment by FernetMenta** [comment]
I did not refresh the browser in time :)

**Comment by mapfau** [comment]
no problem - the only thing is the hack with the mime - type.
Hopefully somebody remembers for what's reason it was implemented.

**Comment by FernetMenta** [comment]
drop that mimetype change from this pr. it is required.

**Comment by mapfau** [comment]
hard to do here because of const ref.
I would better lace it in either 

return CDVDInputStreamFFmpeg(fileitem);
or
return (new CDVDInputStreamFile(fileitem));

(these are the only two classes where it can be used)
Which one really needs it?

**Comment by FernetMenta** [comment]
you can create a temp item for this case

**Comment by mapfau** [comment]
from technical point this is no problem, its from locical point of view.
I think it is wrong at this place!

Look here:

https://github.com/xbmc/xbmc/blob/master/xbmc/cores/paplayer/VideoPlayerCodec.cpp#L94

there is set an mimetype just before createInputStream is called.
Why should factory be allowed to overwrite it? Makes really no sense for me.
Poor PAPlayer wonders why the mimetype doesn't arrive to its destination.

But again my question: is it used in CDVDInputStreamFFmpeg or CDVDInputStreamFile.
I cannot believe that it (the hack) is used in both.

**Comment by mapfau** [comment]
B.t.w: why not using "ContentLokup()" at the place where it is checkt if a HEAD Request is required?
Why the need setting mime types back? If an origin want kodi to let the mime type auto - detect, it should set ContentLookup (and best leave mime types empty)

But what if HEAD isn't supported? Is it better to display no mimetype?
Why not the mime-type wich the origin has set as default?

This should not be forced in the InputStream manager!

**Comment by FernetMenta** [comment]
VideoPlayer is free to detect mimetype for its purposes. Sources may have set it wrong or not at all. A source can prevent player from doing a head request. That's how it works.

> But again my question: is it used in CDVDInputStreamFFmpeg or CDVDInputStreamFile.

I don't understand your question. it is required for selecting one or the other

**Comment by mapfau** [comment]
The latest commit assures that VideoPlayer renews the mime type (if allowed)
Beside this it sets bFolder to true to avoid complex mime type search when directory tree is set up.

**Comment by FernetMenta** [comment]
please slit in smaller junks. the bFolder change is not directly related. note that VideoPlayer is not the only client of DVDInputStream. paplayer is most likely broken now.

**Comment by mapfau** [comment]
ok, will split. There are about 5 places which are affected. paplayer for examle sets mime type exactly before createInputStream is called. Because of that I don't let it reevaluate there

**Comment by FernetMenta** [comment]
> paplayer for examle sets mime type exactly before createInputStream is called

it needs to force HEAD request for those types. just like you did for VideoPlayer

**Comment by mapfau** [comment]
ok, then I'll add it ad al the 5 places.....

**Comment by mapfau** [comment]
were only 3 places where it had to be added, the other were forced to not lookup the mime type using CURL

**Comment by FernetMenta** [comment]
thanks
jenkins build this please

**Comment by FernetMenta** [comment]
@mapfau can you drop ENABLE_DVDINPUTSTREAM_STACK and CDVDInputStreamStack? seems there are still some bits around that make CI unhappy

**Comment by mapfau** [comment]
sure

**Comment by mapfau** [comment]
can you pls. try again?

**Comment by FernetMenta** [comment]
ahh you patched that dead code :) I would have removed it but can be done later
jenkins build this please

**Comment by mapfau** [comment]
sorry, misunderstood it....

**Pull request merged**